integreat_cms/summ_ai_api/utils.py

```diff
@@ -393,3 +418,137 @@ def __repr__(self):
         :rtype: str
         """
         return f"<TranslationHelper (translation: {self.german_translation!r})>"
+
+
+class PatientTaskQueue(deque):
+    """
+    A 'patient' task queue which only hands out sleep tasks after a task was reported as failed.
+    """
+
+    #: The UNIX timestamp when the last rate limited request occurred
+    #:
+    #: :type: float
+    last_rate_limit = None
+
+    #: Seconds to wait after running into the rate limit before sending the next requests
+    #:
+    #: :type: float
+    wait_time = settings.SUMM_AI_RATE_LIMIT_COOLDOWN
+
+    def __init__(
+        self,
+        tasks,
+        wait_time=settings.SUMM_AI_RATE_LIMIT_COOLDOWN,
+    ):
+        """
+        Constructor initializes the class variables
+
+        :param tasks: List of request tasks
+        :type tasks: list [asyncio.Future]
+
+        :param wait_time: Waiting time until start next request in seconds
+        :type wait_time: int
+        """
+        super().__init__(tasks)
+        self.wait_time = wait_time
+
+    def __next__(self):
+        """
+        Extension of the next() function of deque
+        Checks if the queue processing should wait.
+        Ejects the next task or goes to sleep until the end of the waiting time.
+
+        :returns: a task of the queue
+        :rtype: asyncio.Future
+        """
+        if len(self) == 0:
+            raise StopIteration()
+
+        now = time.time()
+        if (
+            self.last_rate_limit is not None
+            and now - self.last_rate_limit < self.wait_time
+        ):
+            # If we are currently waiting out a rate limit,
+            # the worker should sleep for the remaining time before getting the next real task.
+            # Return a function without arguments that the worker can evaluate
+            # to obtain a coroutine.
+            # (see the corresponding comment in SummAiApiClient…translate_text_fields())
+            return partial(asyncio.sleep, self.wait_time - (now - self.last_rate_limit))
+        return self.popleft()
+
```

Comment:
```suggestion
```
Since we're not using the queue as sync iterator anymore, I think we should remove this code block.

Reply:
My question with this is: What is your take on the issue of inheriting from `deque`? I like the idea of having a subclass that behaves like a special form of the base class, which is why I opted to add the async version of `__next__()` in addition to this. I understand why removing `__next__()` instead is a good thing (it's dead code for the current state of the software after all) – but it also breaks the idea that the class does it's special thing but can also be used just as one would use the base class (which in this case doesn't technically break anything but our overarching intent with the class).
It's another vague thing where I have nothing more than two conflicting gut feelings, which don't play nicely with a habit of wanting to do things perfectly in first approximation... :grin:

Reply:
The way I understand it: We're mixing concepts here, since queues are different from iterators, and we need some functionality from both. The `next()` is something defined on iterators, not on queues. So by removing the `__next__` methods, we're not breaking the concept of the base class, it's still a normal queue and you can pop tasks synchronously by using the inbuilt methods `.popleft()` etc... just when you want to use the "patient" feature of the class, then you can use it's async iterator interface...
Theoretically, we could also use [async queues](https://docs.python.org/3/library/asyncio-queue.html), but they are not thread-safe. So we need `dequeue` to get thread-safety, and we need the async iterator stuff to get the queue items asynchronously.
I mean if we really wanted to separate these concepts, we could also define two classes: One queue which contains the tasks, and one iterator which handles the fetching from the queue. But I guess it's also fine in its current form.